heuristic/repair_operators/random_repair.py

```diff
@@ -0,0 +1,35 @@
+from numpy.random import RandomState
+
+from heuristic.classes import Solution
+from heuristic.functions import create_single_customer_route
+
+
+def random_repair(current: Solution, rnd_state: RandomState) -> Solution:
+    """
+    Sequentially inserts a random customer from the unassigned pool in a random
+    feasible location in the Solution.
+
+    Random repair in Hornstra et al. (2020).
+    """
+    rnd_state.shuffle(current.unassigned)
+
+    while len(current.unassigned) != 0:
+        customer = current.unassigned.pop()
+
+        while True:
+            idx_route = rnd_state.randint(len(current.routes) + 1)
+
+            if idx_route < len(current.routes):
+                route = current.routes[idx_route]
+                insert_idx = rnd_state.randint(len(route.customers))
+
+                if route.can_insert(customer, insert_idx, current.problem):
+                    route.insert_customer(customer, insert_idx, current.problem)
+
+                    return current
+
+            else:
+                route = create_single_customer_route(customer, current.problem)
```

Comment:
This is a pretty weak fallback - I'd suggest resampling another route (maybe get a permutation of all routes first, and try them all starting from the first, and only if none are feasible create a single-customer route).

Reply:
If there is infeasibility another sample will be taken, the idx_route drawn can exceed the number of routes (by 1) in which case the else part creates a new route for this customer.

I assumed that an insertion being infeasible would only occur very rarely, in that case checking all feasible options would make the operator much slower than it needs to be. 

The probability that a new route is chosen could of course be lowered.

Reply:
>If there is infeasibility another sample will be taken

Right now you return. That ends the function..

>in that case checking all feasible options

I am not suggesting this. All I am suggesting is you compute a shuffle of the indices for the routes, and then check the routes in that order, stopping the moment you find the first feasible allocation.

Reply:
A new route should be your absolute last resort, so only when no other route is feasible. Which should be a fairly rare event.